### PR TITLE
【OSCP】在 SecretFlow 拆分学习中实现 Cafe 防御

### PR DESCRIPTION
**Pull request by Rainysponge**
**Type of change**

- [ ] Add new papers (Please tell us why you think this paper is awesome!)
- [ ] Fix the category of an existing paper/papers (Please tell us the reasons)
- [x] Add a new tool/primitive/application with a new markdown page (Thank you! Also, please tell us more about this awesome thing!)

**Description**
fixed: #1241
> This PR introduces the [Cafe Fake Gradient Defense](https://openreview.net/pdf?id=b4YiFnQH3gN) into SecretFlow. This algorithm prevents attackers from recovering training data from true gradients by using fake gradients. Specifically, each client generates random fake gradients from a normal distribution, sorts them in descending order, and sorts their true gradients to record the sorted indices. They then compute the L2-norm distance between each true gradient and all fake gradients to find the nearest fake gradient. The fake gradients are paired with true gradients based on the sorted order to keep large/small values in the same positions. Finally, clients upload the fake gradients to the server, ensuring attackers cannot match and invert true inputs.

### CODE REVIEW

File: secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                import torch
+
+                for m in range(M):
+                    _gradient = gradient[m]
+                    Psi = [
+                        torch.normal(mean=0, std=sigma, size=_gradient.shape)
+                        for _ in range(v)
+                    ]
+                    Psi = [torch.sort(g, descending=True)[0] for g in Psi]
+                    _gradient = gradient[m]  # 选取每个客户端的梯度
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py:63]: 和57行重了

File: secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from secretflow.ml.nn.callbacks.callback import Callback
+
+class CAFEFakeGradientsMultiClient(Callback):
+    """
+    TODO: need explanation!
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py:20]: 写点注释解释下算法吧

File: secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                            )
+                            l += 1
+                    fake_gradient.append(_gradient)
+            assert len(fake_gradient) >= 1
+            if len(fake_gradient) == 1:
+                worker._gradient = fake_gradient[0]
+            else:
+                worker._gradient = fake_gradient
+
+        self._workers[self.device_y].apply(
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py:101]: 话说改梯度影响多少模型本身精度

File: tests/ml/nn/sl/defenses/test_cafe_fake_gradients.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        train_label,
+        validation_data=(test_data, test_label),
+        epochs=1,
+        batch_size=128,
+        shuffle=False,
+        random_seed=1234,
+        callbacks=[fake_gradients],
+    )
+    print(history)
+    # assert history["val_BinaryAccuracy"][-1] > 0.6
```
Comment (zhaocaibei123) [inline tests/ml/nn/sl/defenses/test_cafe_fake_gradients.py:105]: 注掉的代码可以删掉

File: secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+from secretflow.ml.nn.callbacks.callback import Callback
+
+class CAFEFakeGradientsMultiClient(Callback):
+    """
+    The method is designed for against CAFE attack: https://arxiv.org/abs/2110.15122.
+    Each local worker randomly generates gradients with the normal distribution N, 
+    sorts them and true gradients in descending order, computes the L2-norm distance
+    to find the nearest fake gradient, pairs fake gradients with true gradients by
+    the sorted order, and uploads the fake gradients to the server.
+    """
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py:25]: 参数注释也加一下吧，参考google风格

File: secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        def fake_gradient(worker, sigma, tua=1.1, M=1, v=128):
+
+            gradient = worker._gradient
+            fake_gradient = []
+            if not isinstance(gradient, list):
+                gradient = [gradient]
+            M = len(gradient)
+
+            if self.backend == "tensorflow":
+                import tensorflow as tf
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py:53]: 可以不引用，这里 raise NotImplementedError()

File: secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        self.backend = backend.lower()
+        self.exec_device = exec_device
+        self.noise_scale = noise_scale
+        self.tua = tua
+        self.v = v
+        super().__init__(**kwargs)
+
+    def on_fuse_backward_end(self):
+        def fake_gradient(worker, sigma, tua=1.1, M=1, v=128):
+
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/defenses/cafe_fake_gradients.py:45]: 用 black 24.2.0，format 下代码

**Comment by Rainysponge** [comment]
这个PR在实现上还有一点问题 等后面我把CAFE加到隐语中再看

**Review by zhaocaibei123 — DISMISSED** [review]

File: secretflow/ml/nn/sl/attacks/cafe_torch.py [hunk 8]
```diff
@@ -0,0 +1,1084 @@
...
+            return _data_x
+
+        def get_real_label(worker):
+            if isinstance(worker.train_y, list):
+                label = worker.train_y[0].cpu().numpy()
+            else:
+                label = worker.train_y.cpu().numpy()
+            return label
+
+        def get_victim_model(victim_worker: SLBaseTorchModel):
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/sl/attacks/cafe_torch.py:207]: 论文的假设是攻击方可以拿到victim的模型和数据吗

Comment (Rainysponge) [reply secretflow/ml/nn/sl/attacks/cafe_torch.py:207]: 不需要数据 但是需要模型参数

**Review by zhaocaibei123 — APPROVED** [review]

**Review by ian-huu — APPROVED** [review]
LGTM
